Add CLI option for disabling tls verification while connecting to s3.

onkarbhat opened the pull request:
Introduce a new CLI option for the kopia command. 
"kopia repistory connect s3 --bucket=`<bucket name>` --disable-tls-verification". 
This new option disables tls verification. 


Testing:
1. Manual tests for connecting to an s3 bucket worked. Snapshot creates, snapshot list, snapshot content commands worked.
2. Updated unit tests.



```
 % kopia repository connect s3 --bucket=onkar-kopia-s3-bucket --disable-tls-verification
Enter password to open repository: 

Connected to repository.

NOTICE: Kopia will check for updates on GitHub every 7 days, starting 24 hours after first use.
To disable this behavior, set environment variable KOPIA_CHECK_FOR_UPDATES=false
Alternatively you can remove the file "/Users/onkarbhat/Library/Application Support/kopia/repository.config.update-info.json".
```


```
% kopia snapshot list
onkarbhat@onkars-mbp:/Users/onkarbhat/kasten-workspace/kopia/kopia/kopia-test-dir
  2020-03-18 13:46:27 PDT k348ab6160c6b64d539c7499109802db5 284 B drwxr-xr-x files:3 dirs:1 (latest-4)
  + 1 identical snapshots until 2020-03-18 13:46:33 PDT
  2020-03-18 13:47:23 PDT k1ed9908f6e20c610f6ed24602ce8dda0 320 B drwxr-xr-x files:4 dirs:1 (latest-2)
  2020-03-18 13:50:06 PDT k385c308ea8023bb5e30569c6dd4a5b49 320 B drwxr-xr-x files:4 dirs:2 (latest-1,annual-1,monthly-1,weekly-1,daily-1,hourly-1)
```


```
% kopia snapshot create kopia-test-dir
Snapshotting onkarbhat@onkars-mbp:/Users/onkarbhat/kasten-workspace/kopia/kopia/kopia-test-dir ...
 * 0 hashing, 0 hashed (0 B), 4 cached (320 B), 2 uploaded (4.3 KB) 100.0%
Created snapshot with root k385c308ea8023bb5e30569c6dd4a5b49 and ID 10931ab1e0dd605bf95fed667aafdcba in 0s
```

ntolia commented:
How did you validate against places with incorrect certificates? I would validate connections at least (with and without the flag) against some of the servers listed on https://badssl.com/.

onkarbhat commented:
@ntolia - I updated the PR with tests for getCustomTransport(). This function is defined in s3_storage.go and used in that file too in New(). 

I added a unit test called TestCustomTransportNoSSLVerify to verify the behavior of InsecureSkipVerify . 

I couldn't think of a way to making the existing minio based tests to test InsecureSkipVerify . If I set the Endpoint to one of the bad ssl links then multiple things will fail  for example createBucket that sets up the test. Even if I use the proper URL for createBucket, and use the bad SSL for testStorage(), then https://github.com/kastenhq/kopia/blob/becf08af8a41f87101c269a19319acf04f55d77f/repo/blob/s3/s3_storage_test.go#L217 will fail at https://github.com/kastenhq/kopia/blob/becf08af8a41f87101c269a19319acf04f55d77f/repo/blob/s3/s3_storage.go#L236  because NewCredentials tries to validate the URL.

ntolia commented:
@onkarbhat Totally get it about the failure but the failure error message should be different. Without `disable-tls-verification`, it should be a TLS issue and with, it should be a credentials issue. That should help this to not regress.

onkarbhat commented on repo/blob/s3/s3_storage_test.go:
```diff
@@ -231,6 +245,32 @@ func testStorage(t *testutil.RetriableT, options *Options) {
+func testURL(url string, t *testing.T) {
+	err := getURL(url, true)
+	if err != nil {
+		t.Fatalf("could not get url:%s, error:%v", url, err)
```
@ntolia - Are you referring to the error strings here and at L270 in ur recent comment ? 

At L 265, since it is the scenario where we disable tls verification, the error string should say it is a credentials issue? (But there are no credentials required for getting the Bad SSL URLs)
And at L270 since we don't disable tls verification, we want the error to be "TLS issue detected" ?

onkarbhat replied:
OR you are referring to the code in testStorage():
https://github.com/kastenhq/kopia/blob/becf08af8a41f87101c269a19319acf04f55d77f/repo/blob/s3/s3_storage_test.go#L222 

We could  add the options.DoNotVerifyTLS in the error message  to help debug failures.

ntolia replied:
@onkarbhat Actually, wasn't referring to any of those particular lines but they might work if it is easy to evaluate. If this is too hard to test, we can move on.

onkarbhat commented:
Addressed review comments.

onkarbhat commented:
One of the bad SSL URLs used in the test didn't meet the test's expectations. So the test failed.

```
--- FAIL: TestCustomTransportNoSSLVerify (0.47s)
    s3_storage_test.go:279: expected a TLS issue, but none found for url:https://incomplete-chain.badssl.com/
```

If you open this URL in a browser , you will see this message:

```
The certificate chain sent by this site is missing an intermediate certificate. This will cause a certificate error unless the browser already has the intermediate certificate in a cache or implements AIA fetching.
```

So there couple of possibilities where accessing this URL will *not* cause a TLS issue on the client side. So I am removing this URL from the list of URLs being tested in this PR.

onkarbhat commented:
@vkamra @ntolia  - I see this error:

```
The base branch restricts merging to authorized users
```
Maybe I have to be added as a contributor ?

ntolia commented:
@onkarbhat Don't worry about merging here. Just open this PR upstream.

julio-lopez commented on repo/blob/s3/op:
```diff
@@ -0,0 +1,201 @@
+2020/03/18 22:28:06 created session token with assume role: expiration: 2020-03-19 05:43:05 +0000 UTC
```
Was this file accidentally included?

onkarbhat replied:
oops yes. My bad. I don't need to update this PR right ? This file did not make it to the upstream PR .

julio-lopez replied:
No need to update this PR.

onkarbhat commented:
@julio-lopez - I did squash all commits before opening the PR upstream. When I addressed Jarek's comments, it resulted in 2 more commits. This is why total number of commits in the upstream PR is 3. I can squash them too in the upstream PR if that is preferable.

julio-lopez commented:
And no need to update the upstream PR. The comment above was more of a guideline
